presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/IntermediateAggregationOptimizer.java

```diff
@@ -0,0 +1,335 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.optimizations;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.SystemSessionProperties;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.DependencyExtractor;
+import com.facebook.presto.sql.planner.Partitioning;
+import com.facebook.presto.sql.planner.PartitioningScheme;
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.ChildReplacer;
+import com.facebook.presto.sql.planner.plan.ExchangeNode;
+import com.facebook.presto.sql.planner.plan.IndexJoinNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.PlanRewriter;
+import com.facebook.presto.sql.planner.plan.SemiJoinNode;
+import com.facebook.presto.sql.planner.plan.TableScanNode;
+import com.facebook.presto.sql.planner.plan.UnionNode;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.ExpressionRewriter;
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.SymbolReference;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+
+import static com.facebook.presto.sql.planner.SystemPartitioningHandle.FIXED_ARBITRARY_DISTRIBUTION;
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.FINAL;
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.INTERMEDIATE;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Scope.REMOTE;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.GATHER;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.REPARTITION;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Inserts an INTERMEDIATE aggregation step with a fixed distribution between a PARTIAL aggregation that
+ * is source partitioned and a FINAL aggregation that occurs on a single node.
+ * <p>
+ * From:
+ * <pre>
+ * - Aggregation (FINAL)
+ *   - Exchange (GATHER)
+ *     - Aggregation (PARTIAL) - same fragment as TableScan
+ * </pre>
+ * to:
+ * <pre>
+ * - Aggregation (FINAL)
+ *   - Exchange (GATHER)
+ *     - Aggregation (INTERMEDIATE)
+ *       - EXCHANGE (ARBITRARY)
+ *         - Aggregation (PARTIAL) - same fragment as TableScan
```

Comment:
@erichwang sorry but I still don't understand motivation behind this `INTERMEDIATE`. Maybe I'm missing something, so I will rephrase what I mean:


Lets consider your proposed plan:

```
- Aggregation (FINAL) - single node  (7)
  - REMOTE Exchange (GATHER)  (6)
    - Aggregation (INTERMEDIATE)  (4, 5)
      - REMOTE EXCHANGE (ARBITRARY)  (3)
        - Aggregation (PARTIAL) - same fragment as TableScan  (1, 2)
          - TableScan 
```

and mine (from 2. of my previous comment):

```
- Aggregation (FINAL)  (7)
  - REMOTE Exchange (GATHER)  (6)
    - Aggregation (PARTIAL)  (1, 2)
      - LOCAL EXCHANGE (ARBITRARY)
        - TableScan
```

Estimation of cost of your plan:

let's assume 1000 nodes cluster and that there are 1000 000 000 input splits, each producing 100 rows (100 billion rows), with `task_concurrency = 16`. Cost of your plan:

1. Cost of consuming raw input: you will create 1000 000 000 of partial aggregations, each will consume 100 rows
2. Cost of producing partial output: each of the 1000 000 000 partial aggregation each will produce one partial output row.
3. Cost of remote distribution: 1000 000 000 partial output rows.
4. Cost of combineing partial outputs: 16 * 1000  Intermediate aggregation operators will have to combine 1000 000 000 partial rows
5. Cost of producing 2nd stage of partial outputs:  16 * 1000 partial output rows
6. Cost of gathering 16 * 1000 partial output rows
7. Cost of Aggregation FINAL, which will have to combine 16 * 1000 partial rows.

Now, cost of my plan:

1. Cost of consuming raw input: 16 * 1000 of partial aggregations, each will consume 100 * 1000 000 000 / (16 * 1000). 
2. Cost of producing partial output: 16 * 1000 partial outputs
3. Cost of remote distribution: 0
4. Cost of combineing partial outputs: 0
5. Cost of producing 2nd stage of partial outputs:  0
6. Cost of gathering 16 * 1000 partial output rows
7. Cost of Aggregation FINAL, which will have to combine 16 * 1000 partial rows.


Costs of 1, 6 and 7 are identical.
Cost of my 2 is the same as yours 5

that means, that your proposal have significantly higher cost, because you have to perform the same work as me, but additionally you have to perform steps 2., 3., 4. This seems to me be true for **ANY** number of nodes, input splits, rows and for **ANY** cost of consuming both raw and partial inputs.